### PR TITLE
Allow oauth decorator to handle custom scopes and add reports:view oa…

### PR DESCRIPTION
**Pull request by sravfeyn**
Allow oauth decorator to handle custom scopes and add reports:view oauth scopes

https://dimagi-dev.atlassian.net/browse/SC-1995

The current default scope `access_apis` gives access to all API endpoints. Ideally, we should restrict each API with specific scopes such as `users`, `apps`, `reports` and `forms` etc ( and further `read`, `write` on each).

But that's beyond the scope of the current work, so I have simply created a new decorator that can handle a custom scope that we are planning to use on SolTech.

### Safety story

The current `api_auth` [decorator](https://github.com/dimagi/commcare-hq/pull/29367) is already rolled out without any issues. The new decorator is only used in one place and I have tested this locally using the API Explorer.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

@czue @proteusvacuum

### CODE REVIEW

File: corehq/apps/domain/decorators.py [hunk 1]
```diff
@@ -341,17 +341,17 @@ def login_or_api_key_ex(allow_cc_users=False, allow_sessions=True, require_domai
     )
 
 
-def login_or_oauth2_ex(allow_cc_users=False, allow_sessions=True, require_domain=True):
+def login_or_oauth2_ex(allow_cc_users=False, allow_sessions=True, require_domain=True, oauth_scopes=['access_apis']):
```
Comment (czue) [inline corehq/apps/domain/decorators.py:344]: nit: mutable default argument. should probably be replaced with a default of `None` and then the default added as the first line of the function:

```
oauth_scopes = oauth_scopes or ['access_apis']
```

File: corehq/apps/domain/decorators.py [hunk 2]
```diff
@@ -341,17 +341,17 @@ def login_or_api_key_ex(allow_cc_users=False, allow_sessions=True, require_domai
...
+        _oauth2_check(oauth_scopes),
         allow_cc_users=allow_cc_users,
         api_key=True,
         allow_sessions=allow_sessions,
         require_domain=require_domain,
     )
 
 
-def get_multi_auth_decorator(default, allow_formplayer=False):
+def get_multi_auth_decorator(default, allow_formplayer=False, oauth_scopes=['access_apis']):
```
Comment (czue) [inline corehq/apps/domain/decorators.py:354]: (same comment)

File: corehq/apps/domain/decorators.py [hunk 3]
```diff
@@ -405,7 +410,7 @@ def wrapped_view(*args, **kwargs):
 basic_auth_or_try_api_key_auth = login_or_basic_or_api_key_ex(allow_sessions=False)
 
 
-def get_auth_decorator_map(allow_cc_users=False, require_domain=True, allow_sessions=True):
+def get_auth_decorator_map(allow_cc_users=False, require_domain=True, allow_sessions=True, oauth_scopes=['access_apis']):
```
Comment (czue) [inline corehq/apps/domain/decorators.py:413]: same

**Review by czue — COMMENTED** [review]
One of the build failures looks real

```
  ======================================================================
  ERROR: corehq.apps.domain.tests.test_auth_decorators:ApiAuthTest.test_api_auth_oauth
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/mnt/commcare-hq-ro/corehq/apps/domain/tests/test_auth_decorators.py", line 216, in test_api_auth_oauth
      self._do_auth_test('bearer myToken', 'corehq.apps.domain.decorators.login_or_oauth2_ex')
    File "/mnt/commcare-hq-ro/corehq/apps/domain/tests/test_auth_decorators.py", line 211, in _do_auth_test
      self.assertEqual(SUCCESS, decorated_view(request, self.domain_name))
    File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 375, in _inner
      function_wrapper = get_auth_decorator_map(allow_cc_users=True, oauth_scopes=oauth_scopes)[authtype]
    File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 424, in get_auth_decorator_map
      OAUTH2: login_or_oauth2_ex(oauth_scopes=oauth_scopes, **decorator_function_kwargs),
  TypeError: mock_auth_decorator() got an unexpected keyword argument 'oauth_scopes'
```

**Review by proteusvacuum — APPROVED** [review]
Once tests pass, and Cory's comments are addressed, this LGTM!

File: corehq/apps/domain/decorators.py (hunk 1 above)
Comment (sravfeyn) [reply corehq/apps/domain/decorators.py]: Ah, that's an easy mistake. Thanks for pointing out.

**Comment by sravfeyn** [comment]
I have fixed the tests (originally missed staging the test changes)

**Review by czue — APPROVED** [review]

**Pull request merged**